pkg/daemon/update.go

```diff
@@ -1597,50 +1589,69 @@ func (dn *Daemon) updateSSHKeys(newUsers []ign3types.PasswdUser) error {
 			concatSSHKeys = concatSSHKeys + string(k) + "\n"
 		}
 	}
-	if !dn.mock {
-		authKeyPath := filepath.Join(coreUserSSHPath, "authorized_keys")
-		authKeyFragmentDirPath := filepath.Join(coreUserSSHPath, "authorized_keys.d")
-
-		if dn.os.IsFCOS() {
-			// In older versions of OKD, the keys were written to `/home/core/.ssh/authorized_keys`.
-			// Newer versions of OKD will however expect the keys at `/home/core/.ssh/authorized_keys.d/ignition`.
-			// Check if the authorized_keys file at the legacy path exists. If it does, remove it.
-			// It will be recreated at the new fragment path by the atomicallyWriteSSHKey function
-			// that is called right after.
-			_, err := os.Stat(authKeyPath)
-			if err == nil {
-				err := os.RemoveAll(authKeyPath)
-				if err != nil {
-					return fmt.Errorf("failed to remove path '%s': %w", authKeyPath, err)
-				}
-			} else if !os.IsNotExist(err) {
-				// This shouldn't ever happen
-				return fmt.Errorf("unexpectedly failed to get info for path '%s': %w", authKeyPath, err)
-			}
 
-			// Ensure authorized_keys.d/ignition is the only fragment that exists
-			keyFragmentsDir, err := ctrlcommon.ReadDir(authKeyFragmentDirPath)
-			if err == nil {
-				for _, fragment := range keyFragmentsDir {
-					if fragment.Name() != "ignition" {
-						keyPath := filepath.Join(authKeyFragmentDirPath, fragment.Name())
-						err := os.RemoveAll(keyPath)
-						if err != nil {
-							return fmt.Errorf("failed to remove path '%s': %w", keyPath, err)
-						}
-					}
-				}
-			} else if !os.IsNotExist(err) {
-				// This shouldn't ever happen
-				return fmt.Errorf("unexpectedly failed to get info for path '%s': %w", authKeyFragmentDirPath, err)
+	authKeyPath := constants.RHCOS8SSHKeyPath
+
+	if !dn.mock {
+		// In RHCOS 8.6 or lower, the keys were written to `/home/core/.ssh/authorized_keys`.
+		// RHCOS 9.0+, FCOS, and SCOS will however expect the keys at `/home/core/.ssh/authorized_keys.d/ignition`.
+		// Check if the authorized_keys file at the legacy path exists. If it does, remove it.
+		// It will be recreated at the new fragment path by the atomicallyWriteSSHKey function
+		// that is called right after.
+		if dn.os.IsEL9() || dn.os.IsSCOS() || dn.os.IsFCOS() {
```

Comment:
Hmm, so the only thing I can think of that could potentially be an issue is that, someone upgrades from RHEL8->9 but they don't have any sshkeys specified, instead, they had previous created the sshkeys via a separate process.

In the old logic we would have just skipped it, but the new EL9() process could wipe any sshkeys that they added via ignition directly or other methods, when we run `cleanSSHKeyPaths` below

But on the other hand, this particular case hopefully is rare enough that it shouldn't be a concern?

Or a third option is if e.g. `/home/core/.ssh/authorized_keys` exists but not otherwise specified via MCO, we add an additional error to let the user know

I could just be over-complicating, let me know if that doesn't make sense

Reply:
To my understanding, the way that the new code path works is it erases the old SSH key path and writes the SSH keys to the new path, effectively recreating it. Other than `$ oc debug node/...`, are there other non-MachineConfig mechanisms an admin could add SSH keys? I only ask because that sounds like a config drift issue.

Speaking of config drift, https://github.com/openshift/machine-config-operator/pull/3538 might cover the case you've mentioned because I have it looking at the old key paths and in those cases, it will degrade. However, maybe it should error if it finds the old key path present (as you've suggested). That PR is dependent upon this one, so would it makes sense to land them together, in addition to the idea you described above?

Reply:
> are there other non-MachineConfig mechanisms an admin could add SSH keys? I only ask because that sounds like a config drift issue.

I think only via ignition directly. I don't think any other operator manages sshkeys.

> Speaking of config drift, https://github.com/openshift/machine-config-operator/pull/3538 might cover the case you've mentioned because I have it looking at the old key paths and in those cases, it will degrade.

Ah I see, I think that makes sense. I am fine landing them in succession, so long as we have something covering for it. I guess it is possible that during the upgrade, the new monitor would take effect and catch any "wrongly placed" sshkeys. That's probably enough to warn the user if anything.

 > in addition to the idea you described above?
It's probably a bit of overkill, I don't off the top of my head know if any users even manage their sshkeys outside of the MCO. I think https://github.com/openshift/machine-config-operator/pull/3538 should be sufficient